packages/plugin-ext/src/main/browser/editor/untitled-resource.ts

```diff
@@ -14,39 +14,91 @@
+import { TextDocumentContentChangeEvent } from 'vscode-languageserver-protocol';
+import { Resource, ResourceResolver } from '@theia/core';
 import URI from '@theia/core/lib/common/uri';
-import { injectable } from 'inversify';
 import { Schemes } from '../../../common/uri-components';
+import { FileSystem } from '@theia/filesystem/lib/common/filesystem';
+import { FileSystemWatcher } from '@theia/filesystem/lib/browser/filesystem-watcher';
+import { OpenerService } from '@theia/core/lib/browser';
+import { FileResource } from '@theia/filesystem/lib/browser';
 
 const resources = new Map<string, UntitledResource>();
 let index = 0;
+
 @injectable()
 export class UntitledResourceResolver implements ResourceResolver {
-    resolve(uri: URI): Resource | Promise<Resource> {
-        if (uri.scheme === Schemes.UNTITLED) {
-            return resources.get(uri.toString())!;
+
+    @inject(FileSystem)
+    protected readonly fileSystem: FileSystem;
+
+    @inject(FileSystemWatcher)
+    protected readonly fileSystemWatcher: FileSystemWatcher;
+
+    @inject(OpenerService)
+    protected readonly openerService: OpenerService;
+
+    async resolve(uri: URI): Promise<UntitledResource> {
+        if (uri.scheme !== Schemes.UNTITLED) {
+            throw new Error('The given uri is not untitled file uri: ' + uri);
+        } else {
+            const untitledResource = resources.get(uri.toString());
+            if (!untitledResource) {
+                return createUntitledResource('', '', uri, this.fileSystem, this.fileSystemWatcher);
+            } else {
+                return untitledResource;
+            }
         }
-        throw new Error(`scheme ${uri.scheme} is not '${Schemes.UNTITLED}'`);
     }
 }
 
 export class UntitledResource implements Resource {
+    private fileResource: FileResource;
 
-    constructor(public uri: URI, private content?: string) {
+    constructor(public uri: URI, private content?: string, private fileSystem?: FileSystem, private fileSystemWatcher?: FileSystemWatcher) {
         resources.set(this.uri.toString(), this);
     }
 
-    readContents(options?: { encoding?: string | undefined; } | undefined): Promise<string> {
-        return Promise.resolve(this.content ? this.content : '');
-    }
-
     dispose(): void {
         resources.delete(this.uri.toString());
     }
+
+    async readContents(options?: { encoding?: string | undefined; } | undefined): Promise<string> {
+        if (this.content) {
+            return Promise.resolve(this.content);
+        } else if (this.fileSystem && await this.fileSystem.exists(this.uri.toString())) {
+            try {
+                const { stat, content } = await this.fileSystem.resolveContent(this.uri.toString(), options);
+                console.debug('The file : ' + this.uri.toString() + ' stat is :' + stat);
+                return content;
+            } catch (e) {
+                throw e;
+            }
+        } else if (this.fileResource) {
```

Comment:
Should not this guard be the first?

Reply:
was fixed

Reply:
i don't get why we should read from `this.content` if we already have `this.fileResource`? If a file was changed on the disk then we should not read some outdated content

Reply:
was fixed. if we have a fileResource then we will read  its content first